### PR TITLE
Improved Calculator Integer keyboard

### LINKED ISSUE #7876: Even more for the Integer keyboard!
@anastasia-mbithe This is looking impressive already.  This is for after the existing pull request has been included.

First 4 minor points:
a) The new-style Help is excellent on the integer keyboard.  But it misses the `gmp` package, at least visibly.  It goes to gmp from the main button, but should also include it as the first in the drop-down list.   Could you please add that.  At the same time - see below, there will be a new section that uses the Zseq package.  Could you add that to the Help too.
b) I have had problems with the hex and oct keys.  This is in data that can't go into hexmode or oxmode.  So could you add 2 more keys, called octmode and hexmode.  They should have the code you currently have in octal and hex.  Then keep the octal and hex keys, but delete the as.octmode and as.hexmode from them.  Then it will give strings for those, just like binary.  Use the same tooltips for the new keys, with one showing the as.hexmode part of the key - as now, and the other omitting that part.

c) Change the tooltip for the bin key slightly, deleting the as.integer from the text.
d) There is also an as.roman function in the R utils package.  (I think R utils is not the same as R.utils?).  So add this key.  This will mean that another row will have 4 keys. The tooltip could be "Converts a small (up to 3899) positive integer to Roman numerals.  For example as.roman(14) = XIV.  

Now could you also arrange for the Zseq package to be added.  This has been improved recently, after I contacted the author.  The commands here are all added in the same way.  It is again up to the length of the data frame, so the nrow bit needs to be included.

Then I am adding proposed 12 keys - so a new section of 3 rows, below the primes - possibly as follows:
1. `palindrome`.  This key generates: `Zseq::Palindromic(n=nrow(x=(data)),gmp=FALSE)`, for a data frame called data. 
If there are 100 rows, this generates the first 100 palindromic numbers.
Tooltip "Generate palindromes to the length of the dataframe.  For example the 100th palindromic number is 909"
2. `fibonnaci2`. Tooltip: "Generates fibonacci numbers to the length of the data frame.  For example the 10th fibonacci number is just 34, but the 50th is much larger, 7778742049". 
The key may have to be: `Zseq::Fibonacci(n=nrow(x=(data)),gmp=TRUE)`.  I think the default is FALSE, though the documentation says otherwise.
3. `abundant`.  Key gives `Zseq::Abundant(n=nrow(x=(data)),gmp=FALSE)`. Tooltip: When divisors sum to more than the number.  So 12 is less than 1+2+3+4+6 = 16.  Opposite is Deficient.
4. `perfect`.  It takes ages if n>10 so key here gives rep(Zseq::Perfect(10),length=nrow(x=(data))) Tooltip: Divisors add to the number.  So 6 = 1+2+3.  Only 51 perfect numbers are currently known.
Now to the second row:
5. `frugal`.  Key gives `Zseq::Frugal(n=nrow(x=(data)),gmp=FALSE)`  125 is frugal because 125 = 5(cubed) and 5 and 3 are fewer digits (2) than 125 (3).
6. `powerful`. Key gives `Zseq::Powerful(n=nrow(x=(data)),gmp=FALSE)`.  36 = 2(squared)*3(squared) is powerful because for each divisor, here 2 and 3, its square is also a divisor.  
7.  `ugly`.  Tooltip "Also called Regular or Hamming or 5-smooth numbers.  Key gives `rep(Zseq::Regular(100),length=nrow(x=(data)),gmp=FALSE)`.  Tooltip is: Numbers for which the factors are only 2, or 3 or 5. 
8. `happy` ends the second row. Key gives `Zseq::Happy(n=nrow(x=(data)),gmp=FALSE)`.  Tooltip: 13 is happy because 1(squared) + 3(squared) = 10, then 1(squared)+ 0(squared) = 1. 
Now the third row:
9. `achilles` starts the 3rd row.  key gives  `Zseq::Achilles(n=nrow(x=(data)))`. Tooltip: Powerful numbers that are not perfect squares.  72 is the smallest Achilles number.
10. `padovan` key gives `Zseq::Padovan(n=nrow(x=(data)))` Tooltip: Sum of last but 1 and last but 2 values. So from ...7, 9, 12, next is 7+9 = 16.
11. `triangle`  key gives  `Zseq::Triangular(n=nrow(x=(data)))`  Tooltip: Number of objects in a triangle, so 0, 1, 3, 6, 10...
12. `square`key gives `Zseq::Square(n=nrow(x=(data)),gmp=FALSE)`  Squares if each integer, so 1, 4, 9, 16.

Finally, in the main keyboard, there is a factorize key which uses the Desctools package.  There is also a factorize key in the gmp package.  Please could that be added - close to the factorize key.  It could be called factorize2.  Tooltip is the same as for factorize.

### PR DESCRIPTION
**Pull request by anastasia-mbithe**
Fixes Partially #7876 

@rdstern, This is ready for testing.

### CODE REVIEW

**Review by rdstern — CHANGES_REQUESTED** [review]
@anastasia-mbithe these improvements are exciting and impressive.  Very well done indeed.
I only have a few very minor points.
a) for fibonacci2 and padovan please omit the gmp=FALSE.  Check they work fine without that, i.e.  just `Zseq::Fibonacci(n=nrow(x=data1))`, and `Zseq::Padovan(n=nrow(x=data1))` , because I think TRUE is the default.
b) For ugly please check the command you give.  it is currently 
`rep(x=Zseq::Regular(n=100), length=nrow(x=data1), gmp=FALSE)`
that looks to me as though the gmp is part of the rep and not the Regular command.  Should it be `rep(x=Zseq::Regular(n=100,gmp=FALSE), length=nrow(x=data1))`
If so, then the same change is needed for Perfect.

These are minor changes and perhaps @lloyddewit could check the code, in parallel, so it can be merged quickly.

**Comment by anastasia-mbithe** [comment]
@rdstern, I have made the changes. For, Perfect and ugly I noticed the default is gmp=TRUE, so I omitted the gmp=FALSE.

**Review by rdstern — CHANGES_REQUESTED** [review]
@anastasia-mbithe just one tiny change - in ugly they remain quite small so we do want to keep the gmp=FALSE there.

**Comment by N-thony** [comment]
@Vitalis95 can you peer review this?

**Review by Vitalis95 — APPROVED** [review]

File: instat/ucrCalculator.vb [hunk 1]
```diff
@@ -217,15 +218,18 @@ Public Class ucrCalculator
         ttCalculator.SetToolTip(cmdFactorial, "factorial n!, as big integer. For example, factorialZ(6)= 720")
         ttCalculator.SetToolTip(cmdChoosez, "computes binomial coefficient choose(n,k) as a big integer. For example, chooseZ(20,2)=190")
         ttCalculator.SetToolTip(cmdNextPrime, "gives the next prime number. For example, nextprime(14)= 17")
-        ttCalculator.SetToolTip(cmdFactorize, "computes the prime factorizations. For example, prime_factors(20)= (2,5,2,1), Factorize(8)= 2:3 for (2,2,2)")
+        ttCalculator.SetToolTip(cmdFactorize, "computes the prime factorizations. For example, prime_factors(20)= (2,5,2,1)")
+        ttCalculator.SetToolTip(cmdFactorize2, "Compute the prime factorization(s) of integer(s) n. For example, Factorize(21)=c(3, 7, 1, 1)")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:222]: make consistent with other tool tip messages?
```suggestion
        ttCalculator.SetToolTip(cmdFactorize2, "computes the prime factorization(s) of integer(s) n. For example, Factorize(21)=c(3, 7, 1, 1)")
```

File: instat/ucrCalculator.vb [hunk 2]
```diff
@@ -217,15 +218,18 @@ Public Class ucrCalculator
...
         ttCalculator.SetToolTip(cmdDivisors, "returns the divisors of x. For example, Divisors(21)= c(1,3,7)")
         ttCalculator.SetToolTip(cmdRankPercent, "returns the percentile that the number correspods to. For example, PercentRank(c(1,2,5,11,15)) = 0.2,0.4,0.6,0.8,1.0")
         ttCalculator.SetToolTip(cmdDigitSum, "calculates digit sum of x. For example, DigitSum(12344)= 14")
-        ttCalculator.SetToolTip(cmdBinary, "converts an integer into a binary number. For example, as.integer(intToBin(c(2,5,7,8)))= 10,101,111,1000")
-        ttCalculator.SetToolTip(cmdAsOctmode, "converts an integer into a octal number. For example, as.octmode(intToOct(c(2,5,12,17)))= 02,05,14,21")
-        ttCalculator.SetToolTip(cmdAsHexmode, "converts an integer into a hexadecimal number. For example, as.hexmode(intToHex(c(2,7,10,15)))= 2,7,a,f")
+        ttCalculator.SetToolTip(cmdBinary, "converts an integer into a binary number. For example,intToBin(c(2,5,7,8))= 10,101,111,1000")
+        ttCalculator.SetToolTip(cmdOctal, "converts an integer into octal number. For example, intToOct(c(2,5,12,17))= 02,05,14,21")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:229]:
```suggestion
        ttCalculator.SetToolTip(cmdOctal, "converts an integer into an octal number. For example, intToOct(c(2,5,12,17))= 02,05,14,21")
```

File: instat/ucrCalculator.vb [hunk 3]
```diff
@@ -217,15 +218,18 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdHex, "converts an integer into hexadecimal number. For example, intToHex(c(2,7,10,15)))= 2,7,a,f")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:230]:
```suggestion
        ttCalculator.SetToolTip(cmdHex, "converts an integer into a hexadecimal number. For example, intToHex(c(2,7,10,15)))= 2,7,a,f")
```

File: instat/ucrCalculator.vb [hunk 4]
```diff
@@ -217,15 +218,18 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdOctmode, "converts an integer into a octal number. For example, as.octmode(intToOct(c(2,5,12,17)))= 02,05,14,21")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:231]:
```suggestion
        ttCalculator.SetToolTip(cmdOctmode, "converts an integer into an octal number. For example, as.octmode(intToOct(c(2,5,12,17)))= 02,05,14,21")
```

File: instat/ucrCalculator.vb [hunk 5]
```diff
@@ -238,6 +242,19 @@ Public Class ucrCalculator
         ttCalculator.SetToolTip(cmdThird, "Third cousin primes, for example third(0,100) gives (89,97)")
         ttCalculator.SetToolTip(cmdTriplets, "Sexy prime triplets, for example triplets(0,100) gives (47, 53, 59)")
         ttCalculator.SetToolTip(cmdKTuple, "k_tuple general formula for all these keys. For example k_tuple(0, 10, c(0,2)) gives twin primes")
+        ttCalculator.SetToolTip(cmdRoman, "converts a small(upto 3899) positive integer to Roman numbers. For example as.roman(14)=XIV")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:245]:
```suggestion
        ttCalculator.SetToolTip(cmdRoman, "converts a small (up to 3899) positive integer to Roman numbers. For example as.roman(14)=XIV")
```

File: instat/ucrCalculator.vb [hunk 6]
```diff
@@ -238,6 +242,19 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdPalindrome, "Generate palindromes to the length of the dataframe. For example the 100th palindromic number is 909")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:246]:
```suggestion
        ttCalculator.SetToolTip(cmdPalindrome, "Generates palindromes to the length of the dataframe. For example the 100th palindromic number is 909")
```

File: instat/ucrCalculator.vb [hunk 7]
```diff
@@ -238,6 +242,19 @@ Public Class ucrCalculator
...
+        ttCalculator.SetToolTip(cmdPerfect, "Divisors add to the number. So 6 = 1+2+3. Only 51 perfect numbers are currently known.")
+        ttCalculator.SetToolTip(cmdFrugal, "125 is frugal because 125 = 5(cubed) and 5 and 3 are fewer digits (2) than 125 (3).")
+        ttCalculator.SetToolTip(cmdPowerful, "36 = 2(squared)*3(squared) is powerful because for each divisor, here 2 and 3, its square is also a divisor.")
+        ttCalculator.SetToolTip(cmdUgly, "Also called Regular or Hamming or 5-smooth numbers. Numbers for which the factors are only 2, or 3 or 5.")
+        ttCalculator.SetToolTip(cmdHappy, "13 is happy because 1(squared) + 3(squared) = 10, then 1(squared)+ 0(squared) = 1.")
+        ttCalculator.SetToolTip(cmdAchilles, "Powerful numbers that are not perfect squares. 72 is the smallest Achilles number.")
+        ttCalculator.SetToolTip(cmdPadovan, "Sum of last but 1 and last but 2 values. So from ...7, 9, 12, next is 7+9 = 16.")
+        ttCalculator.SetToolTip(cmdTriangle, " Number of objects in a triangle, so 0, 1, 3, 6, 10...")
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:256]:
```suggestion
        ttCalculator.SetToolTip(cmdTriangle, "Number of objects in a triangle, so 0, 1, 3, 6, 10...")
```

File: instat/ucrCalculator.vb [hunk 8]
```diff
@@ -3226,4 +3243,165 @@ Public Class ucrCalculator
...
+        clsDataFunction.AddParameter("x", ucrSelectorForCalculations.ucrAvailableDataFrames.cboAvailableDataFrames.SelectedItem)
+
+        clsZseqFunction.SetPackageName("Zseq")
+        clsZseqFunction.SetRCommand(strRCommand)
+        clsZseqFunction.AddParameter("n", clsRFunctionParameter:=clsDataFunction, iPosition:=0)
+        clsZseqFunction.AddParameter("gmp", "FALSE", iPosition:=1)
+
+        ucrReceiverForCalculation.AddToReceiverAtCursorPosition(clsZseqFunction.ToScript, 0)
+    End Sub
+    Private Sub cmdPalindrome_Click(sender As Object, e As EventArgs) Handles cmdPalindrome.Click
```
Comment (lloyddewit) [inline instat/ucrCalculator.vb:3267]:
```suggestion
    End Sub
    
    Private Sub cmdPalindrome_Click(sender As Object, e As EventArgs) Handles cmdPalindrome.Click
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@anastasia-mbithe thanks, just some very small comments

**Review by rdstern — APPROVED** [review]
@lloyddewit this is already a big improvement on the existing Integer keyboard.
I have changed the text at the top to add (partially), so the issue is not closed.  I would like to merge this as it is and then make further progress.  Is the code ok for that?

**Review by lloyddewit — APPROVED** [review]
@anastasia-mbithe Thanks, code looks good

**Pull request merged**